proof/refine/ARM/Ipc_R.thy

```diff
@@ -3161,6 +3161,150 @@ lemma maybeReturnSc_corres:
+  by (wp threadSet_valid_tcbs') (simp add: valid_tcb'_def valid_tcbs'_def tcb_cte_cases_def)
+
+lemma update_valid_tcb[simp]:
+  "valid_tcb ptr tcb (s\<lparr>release_queue := a\<rparr>) = valid_tcb ptr tcb s"
+  "valid_tcb ptr tcb (s\<lparr>reprogram_timer := b\<rparr>) = valid_tcb ptr tcb s"
+  by (auto simp: valid_tcb_def valid_tcb_state_def valid_bound_obj_def valid_arch_tcb_def
+          split: Structures_A.thread_state.splits option.splits)
+
+lemma update_valid_tcbs[simp]:
+  "valid_tcbs (s\<lparr>release_queue := a\<rparr>) = valid_tcbs s"
+  "valid_tcbs (s\<lparr>reprogram_timer := b\<rparr>) = valid_tcbs s"
+  by (simp_all add: valid_tcbs_def)
+
+crunches tcb_release_remove
+  for valid_tcbs[wp]: valid_tcbs
+
+lemma update_valid_tcb'[simp]:
+  "valid_tcb' tcb (s\<lparr>ksReleaseQueue := a\<rparr>) = valid_tcb' tcb s"
+  "valid_tcb' tcb (s\<lparr>ksReprogramTimer := b\<rparr>) = valid_tcb' tcb s"
+  by (auto simp: valid_tcb'_def valid_tcb_state'_def valid_bound_obj'_def
+          split: option.splits thread_state.splits)
+
+lemma update_valid_tcbs'[simp]:
+  "valid_tcbs' (s\<lparr>ksReleaseQueue := a\<rparr>) = valid_tcbs' s"
+  "valid_tcbs' (s\<lparr>ksReprogramTimer := b\<rparr>) = valid_tcbs' s"
+  by (simp_all add: valid_tcbs'_def)
+
+lemma update_tcbInReleaseQueue_False_valid_tcb'[simp]:
+  "valid_tcb' (tcbInReleaseQueue_update (\<lambda>_. False) tcb) s = valid_tcb' tcb s"
+  by (auto simp: valid_tcb'_def tcb_cte_cases_def)
+
+crunches setReleaseQueue, setReprogramTimer
+  for valid_tcb'[wp]: "valid_tcb' tcb"
+  and valid_tcbs'[wp]: "valid_tcbs'"
+
+lemma tcbReleaseRemove_valid_tcbs'[wp]:
+  "tcbReleaseRemove param_a \<lbrace>valid_tcbs'\<rbrace>"
+  by (wpsimp wp: threadSet_valid_tcbs' simp: tcbReleaseRemove_def)
+
+lemma thread_set_empty_tcb_sched_context_weaker_valid_sched_action[wp]:
+  "thread_set (tcb_sched_context_update Map.empty) tcbPtr \<lbrace>weaker_valid_sched_action\<rbrace>"
+  apply (simp only: thread_set_def)
+  apply (wpsimp wp: set_object_wp)
+  apply (clarsimp simp: weaker_valid_sched_action_def pred_tcb_at_def)
+  apply (auto simp: is_tcb_def get_tcb_def obj_at_def  map_project_def tcbs_of_kh_def opt_map_def
+                    pred_map_def map_join_def tcb_scps_of_tcbs_def sc_refill_cfgs_of_scs_def
+             split: option.splits Structures_A.kernel_object.splits)
+  done
+
+lemma schedContextDonate_corres_helper:
+  "(case rv' of SwitchToThread x \<Rightarrow> when (x = t \<or> t = cur) rescheduleRequired
+                             | _ \<Rightarrow> when (t = cur) rescheduleRequired) =
+   (when (t = cur \<or> (case rv' of SwitchToThread x \<Rightarrow> t = x | _ \<Rightarrow> False)) rescheduleRequired)"
+  by (case_tac rv'; clarsimp simp: when_def)
+
+lemma schedContextDonate_corres:
+  "corres dc (invs and sc_at scp and tcb_at thread and weak_valid_sched_action) invs'
+             (sched_context_donate scp thread)
+             (schedContextDonate scp thread)"
```

Comment:
Can we have only the invariants that we really need in the preconditions, instead of having `invs` and `invs’`?

Reply:
Yes, this is done in the following commit.